src/lazy/expanded/compiler.rs

```diff
@@ -631,61 +626,84 @@ impl TemplateCompiler {
 
     /// Given a `LazyValue` that represents a macro ID (name or address), attempts to resolve the
     /// ID to a macro address.
-    fn name_and_address_from_id_expr<'top, D: Decoder>(
-        context: EncodingContextRef<'top>,
-        id_expr: Option<IonResult<LazyValue<'top, D>>>,
+    fn name_and_address_from_id_expr<D: Decoder>(
+        tdl_context: TdlContext,
+        id_expr: Option<IonResult<LazyValue<D>>>,
     ) -> IonResult<(Option<String>, usize)> {
-        match id_expr {
-            None => IonResult::decoding_error("found an empty s-expression in an unquoted context"),
-            Some(Err(e)) => Err(e),
+        // Get the name or address from the `Option<IonResult<LazyValue<_>>>` if possible, or
+        // surface an appropriate error message.
+        let macro_id = match id_expr {
+            None => {
+                return IonResult::decoding_error(
+                    "found an empty s-expression in an unquoted context",
+                )
+            }
+            Some(Err(e)) => return Err(e),
             Some(Ok(value)) => match value.read()? {
                 ValueRef::Symbol(s) => {
                     if let Some(name) = s.text() {
-                        let address =
-                            context.macro_table.address_for_name(name).ok_or_else(|| {
-                                IonError::decoding_error(format!("unrecognized macro name: {name}"))
-                            })?;
-                        Ok((Some(name.to_string()), address))
+                        MacroIdRef::LocalName(name)
                     } else {
-                        IonResult::decoding_error("macro names must be an identifier")
+                        return IonResult::decoding_error("macro names must be an identifier");
```

Comment:
🗺️ Rather than performing the actual lookup of the ID here, this method now saves the ID in a variable (`macro_id` on line 635 above) and then does the lookup on line 665 below.

I believe that we're going to want to eventually support depending on macros in the active macro table; when we go to add that, we'll just add another `.or_else(|| Self::resolve_macro_id_in(...))` call after line 665.